### PR TITLE
Improvements for SPI-based LEDs

### PR DESCRIPTION
**Pull request by embedded-creations**
Still a work in progress...

This code has had limited testing (ESP32 and APA102 LEDs only) on another (private) branch, but not after the commits were cherry picked into this new branch off of dev

To set clkspeed, manually edit cfg.json.  To set the SPI LED order, reverse the fields in LED settings, or manually edit cfg.json

### CODE REVIEW

**Comment by Aircoookie** [comment]
Thank you very much! I see this does not yet actually use the clkspeed value in Neopixelbus - I assume you'd like to add it only once https://github.com/Makuna/NeoPixelBus/pull/441 is merged :) Doing a quick review now...

File: wled00/bus_manager.h [hunk 1]
```diff
@@ -17,11 +17,12 @@ struct BusConfig {
   uint16_t start = 0;
   uint8_t colorOrder = COL_ORDER_GRB;
   bool reversed = false;
+  uint16_t clkspeed;
   uint8_t pins[5] = {LEDPIN, 255, 255, 255, 255};
-  BusConfig(uint8_t busType, uint8_t* ppins, uint16_t pstart, uint16_t len = 1, uint8_t pcolorOrder = COL_ORDER_GRB, bool rev = false) {
-    type = busType; count = len; start = pstart; colorOrder = pcolorOrder; reversed = rev;
+  BusConfig(uint8_t busType, uint8_t* ppins, uint16_t pstart, uint16_t len = 1, uint8_t pcolorOrder = COL_ORDER_GRB, bool rev = false, uint16_t clkSpeed = 5001UL) {
```
Comment (Aircoookie) [inline wled00/bus_manager.h:22]: Why is it `5001UL` instead of `5000UL` here?

File: wled00/bus_manager.h [hunk 2]
```diff
@@ -92,19 +93,21 @@ class BusDigital : public Bus {
...
   BusDigital(BusConfig &bc, uint8_t nr) : Bus(bc.type, bc.start) {
     if (!IS_DIGITAL(bc.type) || !bc.count) return;
-    _pins[0] = bc.pins[0];
-    if (!pinManager.allocatePin(_pins[0])) return;
-    if (IS_2PIN(bc.type)) {
-      _pins[1] = bc.pins[1];
-      if (!pinManager.allocatePin(_pins[1])) {
+    for(int i=0; i<NUM_PINS(bc.type); i++) {
+      _pins[i] = bc.pins[i];
+      Serial.print(" ");
```
Comment (Aircoookie) [inline wled00/bus_manager.h:98]: Some debug isn't bad here, but it shouldn't be plain Serial

File: wled00/bus_manager.h [hunk 3]
```diff
@@ -168,8 +179,12 @@ class BusDigital : public Bus {
     _iType = I_NONE;
     _valid = false;
     _busPtr = nullptr;
+    _clkSpeed = 5000UL;
     pinManager.deallocatePin(_pins[0]);
```
Comment (Aircoookie) [inline wled00/bus_manager.h:183]: seems like a small for would be neater

Comment (Aircoookie) [reply wled00/bus_manager.h:183]: sorry, wrong line. Referring to 183-187.

File: wled00/bus_manager.h (hunk 1 above)
Comment (embedded-creations) [reply wled00/bus_manager.h]: I used unique values for debugging, then missed this one during cleanup

File: wled00/bus_manager.h (hunk 2 above)
Comment (embedded-creations) [reply wled00/bus_manager.h]: That's what I get from trying to submit a PR well after my brain was done for the night :-)

**Comment by embedded-creations** [comment]
> this does not yet actually use the clkspeed value in Neopixelbus - I assume you'd like to add it only once Makuna/NeoPixelBus#441 is merged

Yes, I have a set of changes locally with a commented out `#define USING_UNRELEASED_NPB_FEATURES` that I didn't think was helpful to commit right now

**Comment by embedded-creations** [comment]
LMK if you want to test the clock change feature and I'll make those commits.  I just verified that this branch compiles and works on my ESP32 driving APA102 LEDs

**Comment by embedded-creations** [comment]
This PR was dependent on NeoPixelBus releasing a new version with the dynamic SPI clock speed functionality, which is now done with version 2.6.2.  The NeoPixelBus SPI DMA PR is still pending, but at least there's a release that supports variable clock speed, so we could integrate this feature whenever you want.

**Comment by embedded-creations** [comment]
NeoPixelBus now includes all the features I need for this PR, but it's going to take me some work to merge in all the latest changes and finish the PR.  Would you be open to accepting this PR before the next release, or should I just wait to do the merging until things settle down?

**Comment by blazoncek** [comment]
I think this is still a worthy PR to keep around and eventually implement.
@embedded-creations could you re-base and make new PR into master branch?

**Comment by embedded-creations** [comment]
Yes, I think this should eventually be merged.  Maybe I'll just submit a new PR in the future instead of trying to fix this one after so many changes.

**Pull request closed without merging**